plutus-core/testlib/PlutusCore/Generators/PIR/Utils.hs

```diff
@@ -0,0 +1,75 @@
+{-# LANGUAGE OverloadedStrings #-}
+
+module PlutusCore.Generators.PIR.Utils where
+
+import Prettyprinter
+
+import Data.Set (Set)
+import Data.Set qualified as Set
+import Data.Set.Lens (setOf)
+import Data.String
+import Test.QuickCheck
+
+import PlutusCore.Default
+import PlutusCore.MkPlc hiding (error)
+import PlutusCore.Name
+import PlutusCore.Pretty
+import PlutusIR
+import PlutusIR.Compiler.Datatype
+import PlutusIR.Subst
+
+import PlutusIR.Core.Instance.Pretty.Readable
+
+-- | Show a `Doc` when a property fails.
+ceDoc :: Testable t => Doc ann -> t -> Property
+ceDoc d = counterexample (render d)
+
+-- | Bind a value to a name in a property so that
+-- it is displayed as a `name = thing` binding if the
+-- property fails.
+letCE :: (PrettyPir a, Testable p)
+      => String
+      -> a
+      -> (a -> p)
+      -> Property
+letCE name x k = ceDoc (fromString name <+> "=" <+> prettyPirReadable x) (k x)
+
+-- | Like `forAllShrink` but displays the bound value as
+-- a named pretty-printed binding like `letCE`
+forAllDoc :: (PrettyPir a, Testable p)
+          => String
+          -> Gen a
+          -> (a -> [a])
+          -> (a -> p)
+          -> Property
+forAllDoc name g shr k =
+  forAllShrinkBlind g shr $ \ x ->
+    ceDoc (fromString name <+> "=" <+> prettyPirReadable x)
+          (k x)
+
+-- | Check that a list of potential counterexamples is empty and display the
+-- list as a QuickCheck counterexample if its not.
+assertNoCounterexamples :: PrettyPir a => [a] -> Property
+assertNoCounterexamples []  = property True
+assertNoCounterexamples bad = ceDoc (prettyPirReadable bad) False
+
+-- * Dealing with fresh names
+
+-- | Freshen a TyName so that it does not equal any of the names in the set.
+freshenTyName :: Set TyName -> TyName -> TyName
```

Comment:
Shouldn't this be with the other freshening functions?

Reply:
Renamed the function to `freshenTyNameWith` and added the following comment:

```
-- Note that this doesn't have any relation to 'freshenTyName'. Both functions change the unique,
-- but do it in incomparably different ways.
```

One runs in `Quote`, the other takes a set, inserts `0` and the original name into it... It's quite complicated, hence no reason to compare them and store them next to each other.